### PR TITLE
WIP: Fix flicker on resize under D2D.

### LINKED ISSUE #1201: Flickering when resizing window in Direct2D
There is a flickering that now occurs when resizing a window under the D2D backend.  This was introduced by #1184.

Before #1184: https://files.gitter.im/AvaloniaUI/Avalonia/2ChJ/0.5.2-build4015-alpha.mp4
After: #1184: https://files.gitter.im/AvaloniaUI/Avalonia/FPU6/0.5.2-build4017-alpha.mp4

### PR DESCRIPTION
**Pull request by grokys**
This PR fixes #1201.

Still getting a "rubber banding" effect that needs to be fixed though. This happens because when  `HWndRenderTarget.CreateDrawingContext` gets called, the underlying render may be resized. If this is the case then we need to re-render the composite, but _we're not informed_ that the render target has been resized, so we're left with whatever was drawn at the previous size, stretched to the new size - until the scene is updated and we redraw.

A few possible solutions:

- Add something like `PixelWidth` and `PixelHeight` to `IRenderTarget` and compare the sizes before and after `CreateDrawingContext`
- Add a new interface; something like `IResizableRenderTarget` which will have something like `Size`, `DesiredSize` and `Resize()`
- We might even want something like `ITopLevelRenderTarget` with the members above and also a `Present` method to explicitly cause the double-buffers to be swapped?

The first is a smaller change, but the second and third make what is happening more clear with regards to the underlying mechanisms I think.

@kekekeks what do you think?

### CODE REVIEW

**Comment by kekekeks** [comment]
I don't quite follow. Why would you need a separate `Present` method? Why render target needs to be explicitly resized? Currently render target works on top of something provided by windowing platform, which usually is a fixed size surface. 

And the thing is that surface size won't be really fixed until you call `CreateDrawingContext` (at least with `IFramebufferPlatformSurface`). So, just like with layers, I'd suggest to provide that information from `IDrawingContext`.

File: src/Avalonia.Visuals/Rendering/DeferredRenderer.cs [hunk 1]
```diff
@@ -185,36 +185,40 @@ private void Render(Scene scene)
...
+                        RenderOverlay(scene, context);
+                        RenderComposite(scene, context);
+                    }
+                    else if (composite)
+                    {
+                        context = context ?? _renderTarget.CreateDrawingContext(this);
+                        RenderComposite(scene, context);
                     }
+
+                    context?.Dispose();
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Rendering/DeferredRenderer.cs]: Add a try/finally block to ensure context disposal

**Comment by grokys** [comment]
> Why would you need a separate Present method?

We don't really, it's just that this is done on disposal of the drawing context at the moment, this would simply expose what is happening behind the scenes a little more clearly when you have double buffering.

> Why render target needs to be explicitly resized?

Because the render target being resized is important - we need to re-render when it happens. Again, it being a side-effect of creating a drawing context is surprising from an API-standpoint.

> So, just like with layers, I'd suggest to provide that information from IDrawingContext.

Sorry, provide what information?

**Comment by kekekeks** [comment]
>Sorry, provide what information?

Size of the render target.

>Because the render target being resized is important - we need to re-render when it happens.

Render target should resize automatically inside windowing/rendering backend. Renderer itself should  get the size from the current DrawingContext when it tries to render the frame.

**Comment by kekekeks** [comment]
BTW, this bug also affects GTK3+Skia. But in this case it constantly renders black screen and only renders contents on resize. Can we get the current fix merged?

**Comment by grokys** [comment]
Yeah feel free to merge this and I can fix the remaining problem in another PR

**Pull request merged**